chain/jsonrpc-primitives/src/types/chunks.rs

```diff
@@ -0,0 +1,140 @@
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum ChunkReference {
+    BlockShardId {
+        block_id: near_primitives::types::BlockId,
+        shard_id: near_primitives::types::ShardId,
+    },
+    ChunkHash {
+        chunk_id: near_primitives::hash::CryptoHash,
+    },
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct RpcChunkRequest {
+    #[serde(flatten)]
+    pub chunk_reference: ChunkReference,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct RpcChunkResponse {
+    #[serde(flatten)]
+    pub chunk_view: near_primitives::views::ChunkView,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum RpcChunkError {
+    #[error("The node reached its limits. Try again later. More details: {0}")]
+    InternalError(String),
+    #[error("Block not found")]
+    UnknownBlock(String),
+    #[error("Block `{0}` is unavailable on the node")]
+    UnavailableBlock(near_primitives::hash::CryptoHash),
+    #[error("Shard id {0} does not exist")]
+    InvalidShardId(u64),
+    #[error("Chunk with hash {0:?} has never been observed on this node")]
+    UnknownChunk(near_primitives::sharding::ChunkHash),
+    // NOTE: Currently, the underlying errors are too broad, and while we tried to handle
+    // expected cases, we cannot statically guarantee that no other errors will be returned
+    // in the future.
+    // TODO #3851: Remove this variant once we can exhaustively match all the underlying errors
+    #[error("It is a bug if you receive this error type, please, report this incident: https://github.com/near/nearcore/issues/new/choose. Details: {0}")]
+    Unreachable(String),
+}
+
+impl From<ChunkReference> for near_client_primitives::types::GetChunk {
+    fn from(chunk_reference: ChunkReference) -> Self {
+        match chunk_reference {
+            ChunkReference::BlockShardId { block_id, shard_id } => match block_id {
+                near_primitives::types::BlockId::Height(height) => Self::Height(height, shard_id),
+                near_primitives::types::BlockId::Hash(block_hash) => {
+                    Self::BlockHash(block_hash.into(), shard_id)
+                }
+            },
+            ChunkReference::ChunkHash { chunk_id } => Self::ChunkHash(chunk_id.into()),
+        }
+    }
+}
+
+impl RpcChunkRequest {
+    pub fn parse(value: Option<Value>) -> Result<Self, crate::errors::RpcParseError> {
+        // Try to parse legacy positioned args and if it fails parse newer named args
+        let chunk_reference = if let Ok((chunk_id,)) =
+            crate::utils::parse_params::<(near_primitives::hash::CryptoHash,)>(value.clone())
+        {
+            ChunkReference::ChunkHash { chunk_id }
+        } else if let Ok((block_id, shard_id)) = crate::utils::parse_params::<(
```

Comment:
I believe it should be a tuple of tuple of block id and shard id. Please, confirm that all the expected variants we discussed work

Reply:
I have checked manually with httpie. I can provide logs on Monday for history and to be 100% sure.

Reply:
We have http based test if recall correctly. Maybe it's worth to be extended with such check, what do you think @frol ?

Reply:
Yes, let's make sure we have tests for all the legacy and the new ways of passing the parameters.

Have a good weekend!